wgsl/index.bs

```diff
@@ -6411,7 +6426,145 @@ Issue: (dneto) Default rounding mode is an implementation choice.  Is that what
 Issue: Check behaviour of the f32 to f16 conversion for numbers just beyond the max normal f16 values.
 I've written what an NVIDIA GPU does.  See https://github.com/google/amber/pull/918 for an executable test case.
 
-# Memory Model TODO # {#memory-model}
+# Memory Model # {#memory-model}
+
+In general, [SHORTNAME] follows the [[!VulkanMemoryModel|Vulkan Memory Model]].
+The remainder of this section describes how [SHORTNAME] programs map to the
+Vulkan Memory Model.
+
+## Memory Operation ## {#memory-operation}
+
+A [=read access|memory read=] occurs when an invocation executes one of the
+following:
+* An evaluation of the [=Load Rule=]
+* Any [[#texture-builtin-functions|texture builtin function]] except:
+    * [[#texturedimensions|textureDimensions]]
+    * [[#texturestore|textureStore]]
+    * [[#texturenumlayers|textureNumLayers]]
+    * [[#texturenumlevels|textureNumLevels]]
+    * [[#texturenumsamples|textureNumSamples]]
+* Any atomic built-in function except [[#atomic-store|atomicStore]]
+
+A [=write access|memory write=] occurs when an invocation executes one of the
+following:
+* An [=assignment statement=]
+* A [[#texturestore|textureStore]] built-in function
+* Any atomic built-in function except [[#atomic-load|atomicLoad]]
+    * [[#atomic-rmw|atomicCompareExchangeWeak]] only performs a write if the
+        second element of the returned result is 1
+
+Memory reads and writes do not occur under any other circumstances.
+
+A [memory
+operation](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-memory-operation)
+accesses exactly the set of [=memory location|locations=] associated with the
+particular [=memory view=] used in the operation.
+For example, a memory read that accesses a [=u32=] from a struct containing
+multiple members, only reads the memory locations associated with that u32
+member.
+
+<div class='example memory locations accessed' heading="Accessing memory locations">
+  <xmp>
+    [[block]] struct S {
+      a : f32;
+      b : u32;
+      c : f32;
+    };
+
+    [[group(0), binding(0)]]
+    var<storage> v : S;
+
+    fn foo() {
+      let x = v.b; // Does not access memory locations for v.a or v.c.
+    }
+  </xmp>
+</div>
+
+## Memory Model Reference ## {#memory-model-reference}
+
+[SHORTNAME] forms the following [memory model
+references](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-references):
+* Each variable in [=storage classes/uniform=], [=storage classes/storage=], and
```

Comment:
so, just to get this clear: the section defines memory reads/writes on, say, `function` storage class, but they don't count as "memory references"?

Reply:
I didn't include them because they can't be used for inter-invocation communication, but we can define them to say they have unique memory locations associated with them (during their lifetimes).